checkov/common/checks_infra/solvers/connections_solvers/connection_exists_solver.py

```diff
@@ -25,45 +27,82 @@ def __init__(
-            if not opposite_vertices:
-                continue
+        if not self.vertices_under_resource_types or not self.vertices_under_connected_resources_types:
+            failed.extend(self.vertices_under_resource_types)
+            failed.extend(self.vertices_under_connected_resources_types)
+            return passed, failed, unknown
+
+        if isinstance(graph_connector, Graph):
+
+            for root_vertex in graph_connector.vs:
+                inverted = False
+                origin_attributes = None
+                destination_attributes_list = []
+                for vertex in graph_connector.dfsiter(root_vertex.index):
+                    resource_type = vertex[CustomAttributes.RESOURCE_TYPE]
+                    attributes = vertex["attr"]
+                    if resource_type in self.resource_types and attributes in self.vertices_under_resource_types:
+                        if not origin_attributes:
+                            origin_attributes = attributes
+                        elif inverted:
+                            destination_attributes_list.append(attributes)
+                    elif resource_type in self.connected_resources_types and attributes in self.vertices_under_connected_resources_types:
+                        if not origin_attributes:
+                            origin_attributes = attributes
+                            inverted = True
+                        else:
+                            destination_attributes_list.append(attributes)
 
-            destination_attributes = graph_connector.nodes(data=True)[v]
-            if destination_attributes in opposite_vertices:
-                if origin_attributes in self.excluded_vertices or destination_attributes in self.excluded_vertices:
-                    failed.extend([origin_attributes, destination_attributes])
-                elif origin_attributes in self.unknown_vertices or destination_attributes in self.unknown_vertices:
-                    unknown.extend([origin_attributes, destination_attributes])
-                else:
-                    passed.extend([origin_attributes, destination_attributes])
-                destination_attributes['connected_node'] = origin_attributes
-                continue
+                if origin_attributes and destination_attributes_list:
+                    for destination_attributes in destination_attributes_list:
+                        if origin_attributes in self.excluded_vertices or destination_attributes in self.excluded_vertices:
+                            failed.extend([origin_attributes, destination_attributes])
+                        elif origin_attributes in self.unknown_vertices or destination_attributes in self.unknown_vertices:
+                            unknown.extend([origin_attributes, destination_attributes])
+                        else:
+                            passed.extend([origin_attributes, destination_attributes])
```

Comment:
it's looks like the same logic as in lines 83-89, can you extract it to a function?

Reply:
sounds good, maybe it also makes sense to create separate functions just for the DFS part for both of them, so it is too overloaded here